tensor2tensor/utils/registry.py

```diff
@@ -151,372 +81,285 @@ def default_object_name(obj):
-  if name not in _RANGED_HPARAMS:
-    raise LookupError("RangedHParams set %s never registered." % name)
-  return _RANGED_HPARAMS[name]
+ProblemSpec = collections.namedtuple(
+    "ProblemSpec", ["base_name", "was_reversed", "was_copy"])
 
 
-def list_ranged_hparams():
-  return list(_RANGED_HPARAMS)
+def parse_problem_name(name):
+  """Determines if problem_name specifies a copy and/or reversal.
 
+  Args:
+    name: str, problem name, possibly with suffixes.
 
-def register_problem(name=None):
-  """Register a Problem. name defaults to cls name snake-cased."""
-
-  def decorator(p_cls, registration_name=None):
-    """Registers & returns p_cls with registration_name or default name."""
-    p_name = registration_name or default_name(p_cls)
-    if p_name in _PROBLEMS and not tf.executing_eagerly():
-      raise LookupError("Problem %s already registered." % p_name)
+  Returns:
+    base_name: A string with the base problem name.
+    was_reversed: A boolean.
+    was_copy: A boolean.
+  """
+  # Recursively strip tags until we reach a base name.
+  if name.endswith("_rev"):
+    base, was_rev, was_copy = parse_problem_name(name[:-4])
+    if was_rev:
+      # duplicate rev
+      raise ValueError(
+          "Invalid problem name %s: multiple '_rev' instances" % name)
+    return ProblemSpec(base, True, was_copy)
+  elif name.endswith("_copy"):
+    base, was_reversed, was_copy = parse_problem_name(name[:-5])
+    if was_copy:
+      raise ValueError(
+          "Invalid problem_name %s: multiple '_copy' instances" % name)
+    return ProblemSpec(base, was_reversed, True)
+  else:
+    return ProblemSpec(name, False, False)
+
+
+def get_problem_name(base_name, was_reversed=False, was_copy=False):
+  name = base_name
+  if was_copy:
+    name = "%s_copy" % name
+  if was_reversed:
+    name = "%s_rev" % name
+  return name
+
+
+def _problem_name_validator(k, v):
+  if parse_problem_name(k).base_name != k:
+    raise KeyError(
+        "Invalid problem name: cannot end in %s or %s" % ("_rev", "_copy"))
 
-    _PROBLEMS[p_name] = p_cls
-    p_cls.name = p_name
-    return p_cls
 
-  # Handle if decorator was used without parens
-  if callable(name):
-    p_cls = name
-    return decorator(p_cls, registration_name=default_name(p_cls))
+def _on_problem_set(k, v):
+  v.name = k
 
-  return lambda p_cls: decorator(p_cls, name)
 
+def _call_value(k, v):
+  return v()
 
-def problem(name):
-  """Retrieve a problem by name."""
 
-  def parse_problem_name(problem_name):
-    """Determines if problem_name specifies a copy and/or reversal.
+def _hparams_value_transformer(key, value):
+  out = value()
+  if out is None:
+    raise TypeError("HParams %s is None. Make sure the registered function "
+                    "returns the HParams object" % key)
+  return out
+
+
+model_registry = Registry("models", on_set=_on_model_set)
+optimizer_registry = Registry(
+    "optimizers",
+    default_key_fn=lambda fn: misc_utils.snakecase_to_camelcase(fn.__name__),
```

Comment:
let's actually rm this. have snakecase as the standard, no special case for optimizers.